WalletWasabi.Tests/UnitTests/WabiSabi/Client/CredentialDependencyTests.cs

```diff
@@ -25,99 +25,45 @@ public async void AsyncDependencyGraphTraversalAsync()
-				// Clear unblocked but waiting requests. Not very efficient
-				// (quadratic complexity), but good enough for demonstration
-				// purposes.
-				foreach (var node in UnblockedRequests())
-				{
-					// FIXME this is a little ugly, how should it look in the real code? seems like we're missing an abstraction
-					var edgesByType = DependencyGraph.CredentialTypes.ToImmutableSortedDictionary(t => t, t => g.OutEdges(node, t));
-					var credentialsToRequest = edgesByType.ToImmutableSortedDictionary(kvp => kvp.Key, kvp => kvp.Value.Select(e => e.Value));
-					var credentialsToPresent = pendingCredentialsToPresent[node].ToImmutableSortedDictionary(kvp => kvp.Key, kvp => kvp.Value.AsEnumerable());
+				var inEdges = Enumerable.Concat(g.InEdges(node, CredentialType.Amount), g.InEdges(node, CredentialType.Vsize));
 
-					Assert.NotEmpty(edgesByType);
+				var credentialTasks = inEdges.Select(edge =>taskCompletionSources[edge].Task);
+				await Task.WhenAll(credentialTasks);
 
-					var task = SimulateRequest(node, credentialsToPresent, credentialsToRequest);
+				var credentialsToPresent = credentialTasks.Select(t => t.Result);
+				Assert.Equal<ulong>(inEdges.Select(e => e.Value).OrderBy(x => x), credentialsToPresent.OrderBy(x => x));
 
-					inFlightRequests.Add((Task: task, Dependencies: edgesByType));
-					Assert.True(pendingCredentialsToPresent.Remove(node));
-				}
-
-				// At this point at least one task must be in progress.
-				Assert.True(inFlightRequests.Count > 0);
+				Assert.False(sent.ContainsKey(node));
+				sent[node] = true;
 
-				// Wait for a response to arrive
-				var i = Task.WaitAny(inFlightRequests.Select(x => x.Task).ToArray(), ct.Token);
-				Assert.InRange(i, 0, inFlightRequests.Count);
-				var entry = inFlightRequests[i];
-				inFlightRequests.RemoveAt(i);
-
-				var issuedCredentials = await entry.Task;
+				await Task.Delay(1 + rng.Next(10));
 
-				// Unblock the requests that depend on the issued credentials from this response
-				foreach ((var credentialType, var edges) in entry.Dependencies)
+				foreach (var edge in Enumerable.Concat(g.OutEdges(node, CredentialType.Amount), g.OutEdges(node, CredentialType.Vsize)))
 				{
-					Assert.Equal(edges.Count(), issuedCredentials[credentialType].Count());
-					foreach ((var credential, var edge) in issuedCredentials[credentialType].Zip(edges))
-					{
-						// Ignore the fact that credential is the same as
-						// edge.Value, it's meant to represent the real thing
-						// since it's returned from the task.
-						Assert.Equal(edge.Value, credential);
-						pendingCredentialsToPresent[edge.To][credentialType].Add(credential);
-					}
+					taskCompletionSources[edge].SetResult(edge.Value);
 				}
 			}
 
-			Assert.Empty(inFlightRequests);
-			Assert.Empty(pendingCredentialsToPresent);
+			var tasks = g.Vertices.Select(SimulateRequest).ToArray();
 
-			Assert.True(g.Vertices.All(v => sent.Contains(v)));
+			Task.WaitAll(tasks, TimeSpan.FromMinutes(2));
```

Comment:
This is a blocking method. You can use `Task.WhenAll` to have it async.

Reply:
i prefer `WhenAll`, but how should cancellation/timeouts be handled?

unlike the bounded loop approach before which also had a timeout, this would mean that only the external timeout behavior would cause it to fail if this were to hang, but since the concurrency logic is in the test itself that is unlikely...

Reply:
It is not a real cancel with Task.WaitAll, tasks are still running in the background but you are not awaiting them anymore. To have a real cancellation you have to pass the token to all of the created tasks.  
Preferably to here: 
https://github.com/zkSNACKs/WalletWasabi/pull/5918/files#diff-2a4386371ee7f8e348a7af33752d6508374f6c14398bdbe5310c6643b31c5f4fL62

Reply:
Do you want me to fix this?

Reply:
@nothingmuch?

Reply:
oops, sorry... let's forget this PR for now, it's not valuable in any way directly, i will give it more love after the real version is sorted just to improve readability of the test but right now i want to focus on the coinjoinclient stuff that it is meant to model